### ISSUE TITLE
Support emitting a constant for certain SetVector hardware intrinsic helpers

### ISSUE THREAD

**Issue by tannergooding** [dotnet/coreclr#17225]
For the case where the `SetVector` helper functions are called with all constant arguments, we should support emitting a constant which can be loaded from memory, rather than emitting a chain of shuffle or insert calls.

**Comment by tannergooding** [comment]
FYI. @CarolEidt, @fiigii

**Comment by fiigii** [comment]
> rather than emitting a chain of shuffle or insert calls.

We may need more scenarios and profiling data to support this statement...

**Comment by tannergooding** [comment]
> We may need more scenarios and profiling data to support this statement...

It may be worth noting that MSVC, GCC, and Clang all do this.

They do it for both `_mm_set1` and `_mm_set`. However, I imagine it is only beneficial for the latter (`_mm_set`), since the former (`_mm_set1`) is generally a single broadcast or shuffle.

I do agree we should do the appropriate profiling/etc, but the native compilers doing it provide a good precedent.

**Comment by 4creators** [comment]
> it is only beneficial for the latter (_mm_set), since the former (_mm_set1) is generally a single broadcast or shuffle

It's beneficial for both and is done by all native compilers. Shuffle is cheap but broadcast mostly not.

**Comment by CarolEidt** [comment]
> I do agree we should do the appropriate profiling/etc, but the native compilers doing it provide a good precedent.

There are many things that native compilers do that the JIT doesn't. We need to justify this kind of work with actual scenarios.

**Comment by tannergooding** [comment]
@CarolEidt, this is basically the same as: https://github.com/dotnet/coreclr/issues/17256

I can come up with a bunch of scenarios showing the bad codegen here and actual numbers if it's needed.

**Comment by fiigii** [comment]
> this is basically the same as: #17256 (just replace = new Vector4(1, 2, 3, 4) with = SetVector128(1, 2, 3, 4))

The literals of `float`/`double` have to be read from memory, but integers can be encoded in the instructions. So, `SetVector128/256` should have different codegen heuristics from `Vector4`.

**Comment by tannergooding** [comment]
> The literals of float/double have to be read from memory

Right, so for `float`/`double`, it should be more efficient to do a single read, when all four values are constant.

> but integers can be encoded in the instructions

Could you elaborate?

None of integer SIMD load instructions that I can find (`movd`, `movq`, `movdqa`, `movdqu`, etc) support immediates. So at best, you will be emitting a `mov reg, imm64`, `pinsrq xmm, reg, imm` per 8-bytes (on new hardware, with more complex code involving shifting on older hardware).

**Comment by fiigii** [comment]
> Could you elaborate?
> None of integer SIMD load instructions that
> you will be emitting a mov reg, imm64

Yes, I meant the codegen of "evaluating the integer constants" rather than the SIMD instruction 😄 

>  it should be more efficient to do a single read, when all four values are constant.

Yes, we have to store each constant vector in a memory slot. If a program has many distinct constant vectors, it would spend much memory. There is tread-off...

**Comment by mikedn** [comment]
> If a program has many distinct constant vectors, it would spend much memory. There is tread-off...

I doubt there can be so many distinct constant vectors in a program that it would matter. The real problem is that each method is getting its own data section and thus its own constants. And even inside the same data section constants are not deduplicated so that the same constant is emitted multiple times. Also add the lack of proper packing into the mix and things start to look a bit risky.

All this already happens today for float/double constants but at least those are much smaller...

**Comment by tannergooding** [comment]
> If a program has many distinct constant vectors, it would spend much memory.

I'm not so sure on this part. It looks like it generally takes more bytes to do the insert/shift code than it does to store the raw bytes and read from memory.

```asm
; This takes ~38 bytes of code, plus 16-bytes of storage
vmovss xmm0, dword ptr [rip+0x00]
vinsertps xmm0, xmm0, dword ptr [reg+0x04], 0x10
vinsertps xmm0, xmm0, dword ptr [reg+0x08], 0x20
vinsertps xmm0, xmm0, dword ptr [reg+0x12], 0x30
```

```asm
; This takes ~8 bytes of code, plus 16-bytes of storage
vmovups xmm0, xmmword ptr [reg+0x00]
```

So, even with "perfect" deduping of float constants (which we don't have), we still only have a code savings of ~2-bytes.

**Comment by mikedn** [comment]
Hmm, right, `SetVector` tends to generate so much code that either way things aren't that great in terms of size. The concerns about data size would be more relevant to `SetAll`, that one generates much less code and yet native compilers tends to also use memory constants for that as well. But then native compilers have the luxury of deduplicating constants...

**Comment by tannergooding** [comment]
I've modified this to track `Vector*.Create` as the `SetVector*` methods no longer exist and were changed to the above instead.

**Comment by CarolEidt** [comment]
It looks like #27909 gives us a scenario to motivate this.

**Comment by mikedn** [comment]
Note that what I did with `Vector4` & co. may be insufficient/too limited. The proper solution might require adding `GT_CNS_SIMD`.

**Comment by damageboy** [comment]
I also have a bunch of constants that could use this.

I managed to these into hand-written ReadOnlySpan<byte> proeperty getters + LoadDquVector256().

While the use of `ReadOnlySpan<byte>` initially looks risky, given that these intrinsics are x86 only anyway, and the constant are obviously x86 only as well, it's an ugliness I'm willing to bear.

This mostly alleviates the pain of Vector256.Create() with constants, which is quite considerable.
